### ISSUE TITLE
Thank you!

### ISSUE THREAD

**Issue by btyukodi** [matt-graham/reversible-rng#1]
Not an issue, just wanted to thank you for sharing this! I'm using it to generate a quenched disordered potential energy landscape. Now I don't have to keep the whole damn sequence in the memory any more :D

P. S. Would you happen to have a similar wrapper for a simple linear congruential png? The 624 integers stored make the generator quite heavy... Thanks!

**Comment by btyukodi** [comment]
P. S. Would you happen to have a similar wrapper for a simple linear congruential png? The 624 integers stored make the generator quite heavy...

**Comment by matt-graham** [comment]
Hi. Thanks for the message, good to know this was helpful for at least one other person!

With regards to a lighter-weight linear congruential generator: I don't know of any libraries that have been wrapped to allow direct use in Python, but [this C header-only implementation of a LCG by Johan Helsing](https://github.com/bobbaluba/rlcg) is fairly minimal and should be quite easy to wrap e.g. using Cython.

**Comment by btyukodi** [comment]
Hi, it's me again. I would have one more question regarding the `reverse()` function. Whenever I call `reverse()` and then `standard_uniform()`, the previous random number is re-generated before actually reversing:
`r = revrng.ReversibleRandomState(42432)`
`r1 = r.standard_uniform()`
`r.reverse()`
`r2 = r.standard_uniform()`

In the above case, `r1=r2`. I would expect that calling `standard_uniform()` from a given state would automatically generate the next or previous number, depending on the current direction. It looks like `reverse()` switches the direction but at the same time it moves the generator one step forward so that the next time I call `standard_uniform()` I get back the value before `reverse()`.

In other words, if there was a `next()` and a `prev()` function, I would expect `prev()` to be equivalent to `reverse()` + `standard_uniform()`. Instead, here it is equivalent to `reverse()` + `standard_uniform()` + `standard_uniform()`. My concern is that if `reverse` occurs frequently, generating random numbers takes almost twice as long as in a single directional case.

What do you think?
Thanks
Botond

**Comment by btyukodi** [comment]
Ok, I figured out how to modify it to give `random_int32()` as I described above. The only unclear part is now the `random_uniform()` function. Why wouldn't you simply use `random_int32() / 2^32` if you wanted a random number between `[0, 1)`? Sorry for the flood of questions...

**Comment by matt-graham** [comment]
> In the above case, `r1=r2`. I would expect that calling `standard_uniform()` from a given state would automatically generate the next or previous number, depending on the current direction. It looks like `reverse()` switches the direction but at the same time it moves the generator one step forward so that the next time I call `standard_uniform()` I get back the value before reverse().
> 
> In other words, if there was a `next()` and a `prev()` function, I would expect `prev()` to be equivalent to `reverse()` + `standard_uniform()`. Instead, here it is equivalent to `reverse()` + `standard_uniform()` + `standard_uniform()`. My concern is that if reverse occurs frequently, generating random numbers takes almost twice as long as in a single directional case.

The behaviour you describe is intentional as for my initial use case it made the most sense (to me!). As implemented if a sequence of random variates are generated using a `ReversibleRandomState` and then the `reverse` method called, iterating backwards through the sequence of calls which produced the original variates will produce the same sequence of drawn values in reverse. This necessarily includes the last draw of the original sequence / first draw of the reversed sequence. In my use case there would be a long sequence of draws and then a single reverse switch to then iterate backwards once through the sequence. I agree the current design would be non-optimal however if you are performing reverses with frequency similar to you are generating random values.

> Ok, I figured out how to modify it to give `random_int32()` as I described above. The only unclear part is now the `random_uniform()` function. Why wouldn't you simply use `random_int32() / 2^32` if you wanted a random number between `[0, 1)`? Sorry for the flood of questions...

The current implementation generates *double*-precision floating point values uniformly distributed on `[0,1)` - dividing a single random 32-bit integer by `2^32` would not cover the full range of double-precision floating point values in `[0,1)`. The implementation I used is taken from that in `RandomKit` as used in e.g. NumPy which also means with equivalent seeds a NumPy `RandomState` and the `ReversibleRandomState` will generate the exact same sequence of standard uniform draws which was useful for testing purposes.

**Comment by btyukodi** [comment]
Hi, it makes sense now! Thanks a lot for the detailed explanation!
